File: packages/react-native/ios/NitroPushAnalytics.swift
import Foundation
import UIKit

/// One JS-shaped analytics event. Wire matches the existing `/api/sdk/events`
/// contract — moving from JS to native must not change the server schema.
struct NlAnalyticsEvent: Codable {
    let eventType: String
    let clientUniqueId: String
    let appVersion: String
    let otaVersion: Double?
    let releaseId: String?
    let platform: String
    let osVersion: String?
    let deviceModel: String?
    let occurredAt: String
}

/// Native equivalent of the deleted JS `createAnalyticsEmitter`. Owns the
/// queue, flush timer, retry/backoff, and `URLSession` call. Lives entirely
/// in Swift so events fire even when the JS thread is asleep, mid-restart,
/// or hasn't loaded yet.
///
/// **Threading.** All queue mutation goes through `serial` so callers (the
/// Nitro bridge, lifecycle observers, the rollback sweep) can hammer
/// `enqueue` from any thread without locking.
final class NlAnalytics {
    private let serverUrl: String
    private let deploymentKey: String
    private let capacity: Int
    private let flushAt: Int
    private let flushIntervalSeconds: TimeInterval

    private let serial = DispatchQueue(label: "com.nitropush.analytics", qos: .utility)
    private let session: URLSession

    private var queue: [NlAnalyticsEvent] = []
    private var flushing = false
    private var backoffMs: Int = 0
    private var flushTimer: DispatchSourceTimer?
    private var stopped = false

    init(
        serverUrl: String,
        deploymentKey: String,
        capacity: Int = 200,
        flushAt: Int = 10,
        flushIntervalSeconds: TimeInterval = 30
    ) {
        // Trim trailing slash so we can append `/api/sdk/events` without a
        // double-slash (some reverse proxies treat them as different paths).
        self.serverUrl = serverUrl.hasSuffix("/") ? String(serverUrl.dropLast()) : serverUrl
        self.deploymentKey = deploymentKey
        self.capacity = capacity
        self.flushAt = flushAt
        self.flushIntervalSeconds = flushIntervalSeconds

        let cfg = URLSessionConfiguration.ephemeral
        cfg.timeoutIntervalForRequest = 15
        cfg.timeoutIntervalForResource = 60
        // Telemetry is best-effort and small; we don't want it riding on the
        // device's metered foreground budget.
        cfg.allowsExpensiveNetworkAccess = true
        cfg.allowsConstrainedNetworkAccess = true
        self.session = URLSession(configuration: cfg)
    }

    func enqueue(_ event: NlAnalyticsEvent) {
        serial.async { [weak self] in
            guard let self = self, !self.stopped else { return }
            self.queue.append(event)
            // Drop oldest at capacity — we'd rather lose old events than
            // grow unbounded waiting on a dead network.
            while self.queue.count > self.capacity {
                self.queue.removeFirst()
            }
            if self.queue.count >= self.flushAt {
                self.flushLocked()
            } else {
                self.scheduleTimerLocked()
            }
        }
    }

    func flush() {
        serial.async { [weak self] in self?.flushLocked() }
    }

    func stop() {
        serial.async { [weak self] in
            self?.stopped = true
            self?.flushTimer?.cancel()
            self?.flushTimer = nil
        }
    }

    // MARK: - Private (must be called on `serial`)

    private func scheduleTimerLocked() {
        if flushTimer != nil { return }
        let timer = DispatchSource.makeTimerSource(queue: serial)
        timer.schedule(deadline: .now() + flushIntervalSeconds)
        timer.setEventHandler { [weak self] in
            self?.flushTimer = nil
            self?.flushLocked()
        }
        timer.resume()
        flushTimer = timer
    }

    private func flushLocked() {
        if flushing || stopped || queue.isEmpty { return }
        let batch = queue
        queue.removeAll()
        flushing = true

        guard let url = URL(string: "\(serverUrl)/api/sdk/events") else {
            // Bad URL — drop the batch rather than retrying forever.
            flushing = false
            return
        }

        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.addValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = Body(deploymentKey: deploymentKey, events: batch)
        do {
            req.httpBody = try JSONEncoder().encode(body)
        } catch {
            flushing = false
            return
        }

        session.dataTask(with: req) { [weak self] _, response, error in
            guard let self = self else { return }
            self.serial.async {
                self.flushing = false
                let ok = error == nil
                    && (response as? HTTPURLResponse).map { (200..<300).contains($0.statusCode) } == true
                if ok {
                    self.backoffMs = 0
                } else {
                    // Re-queue the failed batch at the head, exponential
                    // backoff up to 60s between retries.
                    self.queue.insert(contentsOf: batch, at: 0)
                    while self.queue.count > self.capacity {
                        self.queue.removeFirst()
                    }
                    self.backoffMs = min(max(self.backoffMs * 2, 1_000), 60_000)
                    self.scheduleRetryLocked()
                }
            }
        }.resume()
    }

    private func scheduleRetryLocked() {
        flushTimer?.cancel()
        let timer = DispatchSource.makeTimerSource(queue: serial)
        timer.schedule(deadline: .now() + .milliseconds(backoffMs))
        timer.setEventHandler { [weak self] in
            self?.flushTimer = nil
            self?.flushLocked()
        }
        timer.resume()
        flushTimer = timer
    }

    private struct Body: Encodable {
        let deploymentKey: String
        let events: [NlAnalyticsEvent]
    }
}

/// Helpers shared between `NitroPushSdk` (calls these from configure /
/// download / install) and the rollback sweep so events tag with the
/// same device fields regardless of who fires them.
enum NlAnalyticsContext {
    static func now() -> String {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f.string(from: Date())
    }

    static func osVersion() -> String {
        UIDevice.current.systemVersion
    }

    static func deviceModel() -> String {
        var sys = utsname()
        uname(&sys)
        let mirror = Mirror(reflecting: sys.machine)
        let id = mirror.children.compactMap { c -> String? in
            guard let v = c.value as? Int8, v != 0 else { return nil }
            return String(UnicodeScalar(UInt8(v)))
        }.joined()
        return id.isEmpty ? UIDevice.current.model : id
    }
}

File: packages/react-native/android/src/main/java/com/nitropush/sdk/NitroPushAnalytics.kt
package com.nitropush.sdk

import android.os.Build
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import kotlin.math.max
import kotlin.math.min

/**
 * One JS-shaped analytics event. Wire matches the existing `/api/sdk/events`
 * contract — moving from JS to native must not change the server schema.
 */
internal data class NlAnalyticsEvent(
    val eventType: String,
    val clientUniqueId: String,
    val appVersion: String,
    val otaVersion: Double?,
    val releaseId: String?,
    val platform: String,
    val osVersion: String?,
    val deviceModel: String?,
    val occurredAt: String,
)

/**
 * Native equivalent of the deleted JS `createAnalyticsEmitter`. Owns the
 * queue, flush timer, retry/backoff, and HTTP call. Lives entirely in
 * Kotlin so events fire even when the JS thread is asleep, mid-restart,
 * or hasn't loaded yet.
 *
 * **Threading.** All queue mutation runs on the single-thread executor;
 * callers can hammer [enqueue] from any thread without locking.
 */
internal class NlAnalytics(
    serverUrl: String,
    private val deploymentKey: String,
    private val capacity: Int = 200,
    private val flushAt: Int = 10,
    private val flushIntervalMs: Long = 30_000L,
) {
    // Trim trailing slash so we can append `/api/sdk/events` without a
    // double-slash (some reverse proxies treat them as different paths).
    private val serverUrl: String = serverUrl.trimEnd('/')

    private val executor: ScheduledExecutorService =
        Executors.newSingleThreadScheduledExecutor { r ->
            Thread(r, "nitropush-analytics").apply { isDaemon = true }
        }

    private val queue = ArrayDeque<NlAnalyticsEvent>()
    @Volatile private var flushing = false
    @Volatile private var stopped = false
    private var backoffMs: Long = 0
    private var pendingTimer: ScheduledFuture<*>? = null

    fun enqueue(event: NlAnalyticsEvent) {
        executor.execute {
            if (stopped) return@execute
            queue.addLast(event)
            // Drop oldest at capacity — we'd rather lose old events than
            // grow unbounded waiting on a dead network.
            while (queue.size > capacity) {
                queue.removeFirst()
            }
            if (queue.size >= flushAt) {
                flushLocked()
            } else {
                scheduleTimerLocked(flushIntervalMs)
            }
        }
    }

    fun flush() {
        executor.execute { flushLocked() }
    }

    fun stop() {
        executor.execute {
            stopped = true
            pendingTimer?.cancel(false)
            pendingTimer = null
        }
        executor.shutdown()
    }

    // MARK: - Private (must only be called on `executor`)

    private fun scheduleTimerLocked(delayMs: Long) {
        if (pendingTimer != null) return
        pendingTimer = executor.schedule({
            pendingTimer = null
            flushLocked()
        }, delayMs, TimeUnit.MILLISECONDS)
    }

    private fun flushLocked() {
        if (flushing || stopped || queue.isEmpty()) return
        val batch = queue.toList()
        queue.clear()
        flushing = true

        // Network on the executor's single worker — events are tiny and
        // serialized, no contention worries. The connection itself uses
        // its own thread under the hood; we just block this worker until
        // it returns so backoff math reflects the real RTT.
        val ok = postBatch(batch)
        flushing = false
        if (ok) {
            backoffMs = 0
        } else {
            // Re-queue head with exponential backoff up to 60s.
            for (i in batch.indices.reversed()) {
                queue.addFirst(batch[i])
            }
            while (queue.size > capacity) {
                queue.removeFirst()
            }
            backoffMs = min(max(backoffMs * 2, 1_000L), 60_000L)
            pendingTimer?.cancel(false)
            pendingTimer = null
            scheduleTimerLocked(backoffMs)
        }
    }

    private fun postBatch(batch: List<NlAnalyticsEvent>): Boolean {
        val url = try {
            URL("$serverUrl/api/sdk/events")
        } catch (_: Throwable) {
            // Bad URL — drop the batch rather than retrying forever.
            return true
        }
        val payload = JSONObject().apply {
            put("deploymentKey", deploymentKey)
            put("events", JSONArray().apply { batch.forEach { put(it.toJson()) } })
        }
        val body = payload.toString().toByteArray(Charsets.UTF_8)

        var conn: HttpURLConnection? = null
        return try {
            conn = (url.openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                connectTimeout = 15_000
                readTimeout = 60_000
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
                setFixedLengthStreamingMode(body.size)
            }
            conn.outputStream.use { it.write(body) }
            val code = conn.responseCode
            // Drain so the JVM can pool the connection.
            try { conn.inputStream.use { it.readBytes() } } catch (_: Throwable) {}
            code in 200..299
        } catch (_: Throwable) {
            false
        } finally {
            conn?.disconnect()
        }
    }
}

private fun NlAnalyticsEvent.toJson(): JSONObject {
    val obj = JSONObject()
    obj.put("eventType", eventType)
    obj.put("clientUniqueId", clientUniqueId)
    obj.put("appVersion", appVersion)
    obj.put("platform", platform)
    obj.put("occurredAt", occurredAt)
    if (otaVersion != null) obj.put("otaVersion", otaVersion)
    if (releaseId != null) obj.put("releaseId", releaseId)
    if (osVersion != null) obj.put("osVersion", osVersion)
    if (deviceModel != null) obj.put("deviceModel", deviceModel)
    return obj
}

/**
 * Helpers shared between [NitroPushSdk] and the rollback sweep so events
 * tag with the same device fields regardless of who fires them.
 */
internal object NlAnalyticsContext {
    private val ISO: ThreadLocal<SimpleDateFormat> = ThreadLocal.withInitial {
        SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }
    }

    fun now(): String = ISO.get().format(Date())

    fun osVersion(): String = "Android ${Build.VERSION.RELEASE}"

    fun deviceModel(): String = "${Build.MANUFACTURER} ${Build.MODEL}"
}
